server/src-lib/Hasura/RQL/DDL/Remote/Types.hs

```diff
@@ -0,0 +1,157 @@
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE RecordWildCards #-}
+-- |
+
+module Hasura.RQL.DDL.Remote.Types where
+
+import           Hasura.Prelude
+import           Hasura.RQL.DDL.Remote.Input
+import           Hasura.RQL.Instances ()
+import           Hasura.RQL.Types.Common
+import           Hasura.SQL.Types
+
+import           Data.Aeson as A
+import           Data.Aeson.Casing
+import           Data.Aeson.TH
+import           Data.Aeson.Types (Parser)
+import           Data.HashMap.Strict (HashMap)
+import qualified Data.HashMap.Strict as HM
+import           Data.Scientific
+import           Data.Set (Set)
+import           Data.Text (Text)
+import qualified Database.PG.Query as Q
+import           Instances.TH.Lift ()
+import qualified Language.GraphQL.Draft.Syntax as G
+import           Language.Haskell.TH.Syntax (Lift)
+
+data RemoteField =
+  RemoteField
+    { rmfRemoteRelationship :: !RemoteRelationship
+    , rmfGType :: !G.GType
+    , rmfParamMap :: !(HashMap G.Name InpValInfo)
+    }
+  deriving (Show, Eq, Lift)
+
+data RemoteRelationship =
+  RemoteRelationship
+    { rtrName :: RemoteRelationshipName
+    , rtrTable :: QualifiedTable
+    , rtrRemoteSchema :: RemoteSchemaName
+    , rtrRemoteField :: G.Name
+    , rtrHasuraFields :: Set FieldName
+    , rtrRemoteArguments :: RemoteArguments
+    }
+  deriving (Show, Eq, Lift)
+
+newtype RemoteSchemaName
+  = RemoteSchemaName
+  { unRemoteSchemaName :: Text}
+  deriving (Show, Eq, Lift, Hashable, ToJSON, ToJSONKey, FromJSON, Q.ToPrepArg, Q.FromCol)
+
+newtype RemoteRelationshipName
+  = RemoteRelationshipName
+  { unRemoteRelationshipName :: Text}
+  deriving (Show, Eq, Lift, Hashable, ToJSON, ToJSONKey, FromJSON, Q.ToPrepArg, Q.FromCol)
+
+$(deriveJSON (aesonDrop 3 snakeCase) ''RemoteRelationship)
+
+--------------------------------------------------------------------------------
+-- Custom JSON roundtrip instances for RemoteField and down
+
+instance ToJSON RemoteField where
+  toJSON RemoteField {..} =
+    object
+      [ "remote_relationship" .= toJSON rmfRemoteRelationship
+      , "g_type" .= toJsonGType rmfGType
+      , "param_map" .= fmap toJsonInpValInfo rmfParamMap
+      ]
+
+instance FromJSON RemoteField where
+  parseJSON value = do
+    hmap <- parseJSON value
+    rmfRemoteRelationship <- hmap .: "remote_relationship"
+    rmfGType <- hmap .: "g_type" >>= parseJsonGType
+    rmfParamMap <- hmap .: "param_map" >>= traverse parseJsonInpValInfo
+    pure RemoteField {..}
```

Comment:
Can you tell me about this `{..}` syntax ?

Reply:
It's a record wild card. It means "use the names in scope that correspond to the fields in `RemoteField`". Avoids repetition.